## Heroes of Pymoli

In [1]:
# Dependencies
import pandas as pd

# Loading...
data_file = "Resources/purchase_data.csv"

# Pandas reading data!
purchase_data = pd.read_csv(data_file)

### Player Count

In [2]:
unique_players = len(purchase_data["SN"].value_counts())

pd.DataFrame({
    'Total Players':[unique_players]
})

Total Players
0            576

### Purchasing Analysis (Total)

In [3]:
def format(x):
        return "${:,.2f}".format(x)

unique_items = purchase_data["Item ID"].unique()
average_price = purchase_data["Price"].mean()
total_purchases = purchase_data["Purchase ID"].count()
total_revenue = purchase_data["Price"].sum()


purchasing_analysis = pd.DataFrame({
    "Number of Unique Items":[len(unique_items)],
    "Average Price":[average_price],
    "Number of Purchases":[total_purchases],
    "Total Revenue":[total_revenue],
})

purchasing_analysis["Average Price"] = purchasing_analysis["Average Price"].apply(format)
purchasing_analysis["Total Revenue"] = purchasing_analysis["Total Revenue"].apply(format)

purchasing_analysis.round(2)

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

### Gender Demographics

In [4]:
def format(x):
    return "{:,.2%}".format(x)

unique = pd.DataFrame(purchase_data)
unique = unique.drop_duplicates(subset = ["SN"])

unique_gender_count = unique["Gender"].value_counts()
male_percentage = (unique_gender_count["Male"] / unique_gender_count.sum()) 
female_percentage = (unique_gender_count["Female"] / unique_gender_count.sum()) 
other_percentage = (unique_gender_count["Other / Non-Disclosed"] / unique_gender_count.sum()) 

gender_demo = pd.DataFrame({
    " ": ["Male", "Female", "Other / Non-Disclosed"],
   "Total Count": [unique_gender_count["Male"], unique_gender_count["Female"], unique_gender_count["Other / Non-Disclosed"]],
   "Percentage of Players": [male_percentage, female_percentage, other_percentage]
})

gender_demo["Percentage of Players"] = gender_demo["Percentage of Players"].apply(format)

gender_demo.set_index(" ", inplace = True)
gender_demo.round(2)



Total Count Percentage of Players
                                                        
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

### Purchasing Analysis (Gender)

In [5]:
def format(x):
        return "${:,.2f}".format(x)

purgender = purchase_data["Gender"]
pur_gender_count = purgender.value_counts()
male_purch = purchase_data["Gender"].isin(["Male"])
female_purch = purchase_data["Gender"].isin(["Female"])
other_purch = purchase_data["Gender"].isin(["Other / Non-Disclosed"])
avg_purprice = purchase_data["Price"]
male_ppp = purchase_data[male_purch]["Price"].sum() / len(purchase_data[male_purch]["SN"].unique())
female_ppp = purchase_data[female_purch]["Price"].sum() / len(purchase_data[female_purch]["SN"].unique())
other_ppp = purchase_data[other_purch]["Price"].sum() / len(purchase_data[other_purch]["SN"].unique())


gender_analysis = pd.DataFrame({
    "Gender": ["Male", "Female", "Other / Non-Disclosed"],
    "Purchase Count": [pur_gender_count["Male"], pur_gender_count["Female"], pur_gender_count["Other / Non-Disclosed"]],
    "Average Purchase Price": [purchase_data[male_purch]["Price"].mean(), purchase_data[female_purch]["Price"].mean(), purchase_data[other_purch]["Price"].mean()],
    "Total Purchase Value": [purchase_data[male_purch]["Price"].sum(), purchase_data[female_purch]["Price"].sum(), purchase_data[other_purch]["Price"].sum()],
    "Avg Total Purchase per Person": [male_ppp, female_ppp, other_ppp]
    
})

gender_analysis["Average Purchase Price"] = gender_analysis["Average Purchase Price"].apply(format)
gender_analysis["Total Purchase Value"] = gender_analysis["Total Purchase Value"].apply(format)
gender_analysis["Avg Total Purchase per Person"] = gender_analysis["Avg Total Purchase per Person"].apply(format)

gender_analysis.set_index("Gender", inplace = True)

gender_analysis


Purchase Count Average Purchase Price  \
Gender                                                         
Male                              652                  $3.02   
Female                            113                  $3.20   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Male                             $1,967.64                         $4.07  
Female                             $361.94                         $4.47  
Other / Non-Disclosed               $50.19                         $4.56

### Age Demographics

In [6]:
def format(x):
    return "{:,.2%}".format(x)

unique_age = unique["Age"]

age_10less = unique.loc[unique["Age"] < 10].loc[unique["Age"] > 5]
age_1014 = unique.loc[unique["Age"] >= 10].loc[unique["Age"] < 15]
age_1519 = unique.loc[unique["Age"] >= 15].loc[unique["Age"] < 20]
age_2024 = unique.loc[unique["Age"] >= 20].loc[unique["Age"] < 25]
age_2529 = unique.loc[unique["Age"] >= 25].loc[unique["Age"] < 30]
age_3034 = unique.loc[unique["Age"] >= 30].loc[unique["Age"] < 35]
age_3539 = unique.loc[unique["Age"] >= 35].loc[unique["Age"] < 40]
age_40 = unique.loc[unique["Age"] >= 40]

age_demo = pd.DataFrame({
    " ": ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"],
    "Total Count": [len(age_10less["Age"]), len(age_1014["Age"]), len(age_1519["Age"]), len(age_2024["Age"]), len(age_2529["Age"]), len(age_3034["Age"]), len(age_3539["Age"]), len(age_40["Age"])],
    "Percentage of Players": [len(age_10less["Age"]) / len(unique_age), len(age_1014["Age"]) / len(unique_age), len(age_1519["Age"]) / len(unique_age), len(age_2024["Age"]) / len(unique_age), len(age_2529["Age"]) / len(unique_age), len(age_3034["Age"]) / len(unique_age), len(age_3539["Age"]) / len(unique_age), len(age_40["Age"]) / len(unique_age)],
    
})

age_demo["Percentage of Players"] = age_demo["Percentage of Players"].apply(format)

age_demo.set_index(" ", inplace = True)

age_demo


Total Count Percentage of Players
                                        
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

### Purchasing Analysis (Age)

In [7]:
def format(x):
        return "${:,.2f}".format(x)

purage = purchase_data["Age"]
pur_age_count = purage.value_counts()

purage_10less = purchase_data[purchase_data["Age"] < 10].loc[purchase_data["Age"] > 5]
purage_1014 = purchase_data.loc[purchase_data["Age"] >= 10].loc[purchase_data["Age"] < 15]
purage_1519 = purchase_data.loc[purchase_data["Age"] >= 15].loc[purchase_data["Age"] < 20]
purage_2024 = purchase_data.loc[purchase_data["Age"] >= 20].loc[purchase_data["Age"] < 25]
purage_2529 = purchase_data.loc[purchase_data["Age"] >= 25].loc[purchase_data["Age"] < 30]
purage_3034 = purchase_data.loc[purchase_data["Age"] >= 30].loc[purchase_data["Age"] < 35]
purage_3539 = purchase_data.loc[purchase_data["Age"] >= 35].loc[purchase_data["Age"] < 40]
purage_40 = purchase_data.loc[purchase_data["Age"] >= 40]

ppp_10less = purage_10less["Price"].sum() / len(purage_10less["SN"].unique())
ppp_purage_1014 = purage_1014["Price"].sum() / len(purage_1014["SN"].unique())
ppp_purage_1519 = purage_1519["Price"].sum() / len(purage_1519["SN"].unique())
ppp_purage_2024 = purage_2024["Price"].sum() / len(purage_2024["SN"].unique())
ppp_purage_2529 = purage_2529["Price"].sum() / len(purage_2529["SN"].unique())
ppp_purage_3034 = purage_3034["Price"].sum() / len(purage_3034["SN"].unique())
ppp_purage_3539 = purage_3539["Price"].sum() / len(purage_3539["SN"].unique())
ppp_purage_40 = purage_40["Price"].sum() / len(purage_40["SN"].unique())


purage_analysis = pd.DataFrame({
    "Age Ranges": ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"],
    "Purchase Count": [len(purage_10less["Age"]), len(purage_1014["Age"]), len(purage_1519["Age"]), len(purage_2024["Age"]), len(purage_2529["Age"]), len(purage_3034["Age"]), len(purage_3539["Age"]), len(purage_40["Age"])],
    "Average Purchase Price": [purage_10less["Price"].mean(), purage_1014["Price"].mean(), purage_1519["Price"].mean(), purage_2024["Price"].mean(), purage_2529["Price"].mean(), purage_3034["Price"].mean(), purage_3539["Price"].mean(), purage_40["Price"].mean(),],
    "Total Purchase Value": [purage_10less["Price"].sum(), purage_1014["Price"].sum(), purage_1519["Price"].sum(), purage_2024["Price"].sum(), purage_2529["Price"].sum(), purage_3034["Price"].sum(), purage_3539["Price"].sum(), purage_40["Price"].sum()],
    "Avg Total Purchase per Person": [ppp_10less, ppp_purage_1014, ppp_purage_1519, ppp_purage_2024, ppp_purage_2529, ppp_purage_3034, ppp_purage_3539, ppp_purage_40]
})


purage_analysis.set_index("Age Ranges", inplace = True)

purage_analysis["Average Purchase Price"] = purage_analysis["Average Purchase Price"].apply(format)
purage_analysis["Total Purchase Value"] = purage_analysis["Total Purchase Value"].apply(format)
purage_analysis["Avg Total Purchase per Person"] = purage_analysis["Avg Total Purchase per Person"].apply(format)

purage_analysis

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

### Top Spenders

In [8]:
def format(x):
        return "${:,.2f}".format(x)


In [9]:
average_purchase_price = purchase_data.groupby('SN')['Price'].mean()
total_price_value = purchase_data.groupby('SN')['Price'].sum()
purchase_count = purchase_data.groupby('SN')['Purchase ID'].count()

result_dataframe = pd.DataFrame({
    'Purchase Count':purchase_count,
    'Average Purchase Price':average_purchase_price,
    'Total Purchase Value':total_price_value
})

result_dataframe = result_dataframe.sort_values(by='Total Purchase Value',ascending=False)
result_dataframe['Total Purchase Value'] = result_dataframe['Total Purchase Value'].apply(format)
result_dataframe['Average Purchase Price'] = result_dataframe['Average Purchase Price'].apply(format)

result_dataframe.head()


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

### Most Popular Items

In [10]:
def format(x):
        return "${:,.2f}".format(x)

In [11]:
purchase_data.head()

revised_df = purchase_data[['Item ID', 'Item Name', 'Price']]
result_df = revised_df.groupby(['Item ID', 'Item Name'])['Item ID'].count().to_frame()
total_purchase_value = revised_df.groupby(['Item ID', 'Item Name'])['Price'].sum()
average_item_price = revised_df.groupby(['Item ID', 'Item Name'])['Price'].mean()

result_df = result_df.rename(columns = {'Item ID':'Purchase Count'})
result_df['Item Price'] = average_item_price
result_df['Total Purchase Value'] = total_purchase_value 

result_df = result_df.sort_values(by='Purchase Count',ascending=False)

result_df['Total Purchase Value'] = result_df['Total Purchase Value'].apply(format)
result_df['Item Price'] = result_df['Item Price'].apply(format)

result_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

### Most Profitable Items

In [12]:
purchase_data.head()

revised_df = purchase_data[['Item ID', 'Item Name', 'Price']]
result_df = revised_df.groupby(['Item ID', 'Item Name'])['Item ID'].count().to_frame()
total_purchase_value = revised_df.groupby(['Item ID', 'Item Name'])['Price'].sum()
average_item_price = revised_df.groupby(['Item ID', 'Item Name'])['Price'].mean()

result_df = result_df.rename(columns = {'Item ID':'Purchase Count'})
result_df['Item Price'] = average_item_price
result_df['Total Purchase Value'] = total_purchase_value 

result_df = result_df.sort_values(by='Total Purchase Value',ascending=False)

result_df['Total Purchase Value'] = result_df['Total Purchase Value'].apply(format)
result_df['Item Price'] = result_df['Item Price'].apply(format)

result_df.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## Three Observable Trends

* 1. Players within the age range of 20-24 make up the majority of the player base, and also make the most transactions in-game. While the total purchase value is highest, as this age range makes the most purchases, average purchase per person falls between the median and upper quartile at $4.32.


* 2. Most of the game's player base consists of players that identify as male. Although data on female players is lacking in comparison to data on male players, the data suggests that female players spend more money per in-game purchase than male players. 


* 3. The top 3 most purchased items are all priced around the upper quartile price range. This may indicate that popular items such as "Final Critic" and "Oathbreaker, Last Hope of the Breaking Storm" will sell for a higher price due to item rarity and high demand. 

